[Multistage] Runtime changes for leveraging V1 Aggregation Functions

vvivekiyer opened the pull request:
This work is a collaboration between @somandal and myself.
This PR should also be reviewed with @somandal's  https://github.com/apache/pinot/pull/10846

### Motivation:

- Today only a subset of aggregation functions (sum, min, max, count, avg, count(distinct), kurtosis, and skewness) work in the multistage engine.
- To support each new aggregation function changes are required on both the planner and runtime side. Pinot contributors should not need to understand the nitty gritty details of the planner and execution engine to add new aggregation functions.
- The multistage AggregateOperator essentially reimplements the aggregation logic rather than reusing the aggregation logic present in the v1 engine functions. This results in code duplication and a high maintenance overhead (e.g. bug fixes must be made in two places which may even have different logic or different mechanisms to support things like null handling).
- The v1 execution engine supports a large number of aggregation functions and it is not practical to manually add support for each new aggregation function to multistage. This also results in a situation in which each new aggregation function needs to be added to both the v1 engine and the multistage engine, or if support is only added for the v1 engine, the multistage engine must constantly play catch up.

To get around the above limitations we are refactoring the aggregation function framework to work for both the v1 and multistage engine. This PR tackles the changes required on the **runtime** side to use the existing v1 Aggregation functions. This is phase 1 of the planned set of changed to support the aggregation functions and tackles only the existing set of aggregation functions already supported in the multistage engine.

- OSS Issue: https://github.com/apache/pinot/issues/10745
- Design document: https://docs.google.com/document/d/1Us6aBvTpNLMEy0ODo34OgTk73h_LVFFAH6q17689h1M/edit?usp=sharing
- @somandal's  planner changes to support this: https://github.com/apache/pinot/pull/10846


**Note:** This PR is safe to merge and should be merged before the Planner PR (https://github.com/apache/pinot/pull/10846). The aggregation changes are implemented as a new operator. So this code will not actually be used until the planner PR is merged.

### Steps to merge Planner PR (after this Runtime PR is merged):
- Rebase and resolve conflicts (this will pull in merged runtime changes)
- Delete the existing `AggregateOperator.java`. Rename `NewAggregateOperator.java` to `AggregateOperator.java` (see the commented code in `PhysicalPlanVisitor.visitAggregate()`)
- Look for “TODO(Sonam)” comments across the codebase (called out files below) and address them:
1. AggregationFunctionFactory.java
2. NewAggregatorOperator.java
3. PhysicalPlanVisitor.java
4. AggregateOperatorTest.java
5. Aggregates.json
- Review, run tests and merge code.


### Future Work:

- Today SUM and SUM0 are handled similarly. In the future, we can have separate v1 aggregation functions SUM and SUM0 where the return types are differentiated. Today Calcite rewrites all SUM as SUM0, and if the types are non-nullable then they do not even add a CASE operator to return null for SUM = 0 (which is true for all scenarios we tested based on the planner tests)
- Today, when the AggregationOperator (in leaf stage) executes aggregations like MIN, MAX, SUM on a column where null handling is disabled, it would explicitly return infinity, -infinity and 0 respectively. However, if the same aggregation is handled in the AggregateOperator (in intermediate stage) we would get null as a result. This discrepancy exists today even without our changes.
- Calcite divides AVG aggregation functions vertically into SUM and COUNT. The division operator (/) is then used to perform SUM / COUNT. However, our v1 aggregation code could return 0 as the default value for COUNT and this will result in division by zero error. This could not be handled without a scalar function today because all our aggregation fields are set to non-nullable types according to Calcite. Using a Calcite CASE operator to return null for COUNT=0 and only then perform division results in a planning error because of type mismatch between nullable vs the expected non-nullable type.
- Need to change some of the functions such as SUM, MIN, MAX to use the appropriate types rather than forcing them to return a double result.
- Need to add support to take literal arguments for the aggregation functions (e.g. PERCENTILE, DISTINCTCOUNTHLL, etc) and add tests for all such functions to ensure they work as expected.
- Add support for certain aggregation functions which take more than one column as argument (e.g. COVARIANCE)
- ArgMin/ArgMax support



cc @siddharthteotia @somandal  @walterddr @kishoreg

abhioncbr commented on pinot-core/src/main/java/org/apache/pinot/core/common/IntermediateStageBlockValSet.java:
```diff
@@ -0,0 +1,250 @@
+
+  @Override
+  public String[][] getStringValuesMV() {
+    throw new UnsupportedOperationException();
```
Hi, 
Thanks for putting this work. I was waiting for this as I think it helps me with supporting the [json scalar functions](https://github.com/apache/pinot/pull/10790) in multistage. 

Just a question, why we are not supporting MV types? With the MV types support we can handle scalar functions like `string split` and many other.

vvivekiyer replied:
We currently don't support MV types in Multistage engine - https://github.com/apache/pinot/issues/10658. Eventually when we support MV columns in Multistage and add the corresponding aggregation functions, we'll have to support MV types here.

walterddr commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/PhysicalPlanVisitor.java:
```diff
@@ -96,10 +102,88 @@ public MultiStageOperator visitMailboxSend(MailboxSendNode node, PhysicalPlanCon
   public MultiStageOperator visitAggregate(AggregateNode node, PhysicalPlanContext context) {
     MultiStageOperator nextOperator = node.getInputs().get(0).visit(this, context);
+
+    // TODO: Will need to use a collection of inputSchema when we support aggregation functions with multiple
+    //  columns.
```
i am not quite sure what this TODO is about, could you elaborate with an example?

vvivekiyer replied:
This is a misplaced TODO. I've removed it.

walterddr commented:
thank you for the contribution. @vvivekiyer and @somandal 

could i ask a generic question what's the relationship between this PR and #10846? i guess specifically how do i review them? should i download both, cherry-pick/merge together and try at the same time? please share your thoughts. many thanks!

somandal commented:
> thank you for the contribution. @vvivekiyer and @somandal
> 
> could i ask a generic question what's the relationship between this PR and #10846? i guess specifically how do i review them? should i download both, cherry-pick/merge together and try at the same time? please share your thoughts. many thanks!

@walterddr yes these two PRs should ideally be reviewed together. https://github.com/apache/pinot/pull/10846 is the planner side changes necessary for this feature. Keep in mind that we haven't yet added support for function arguments as literals etc which are heavily used for some of our aggregation functions (and this will be done as part of phase 2). 
Merging together may not be enough, since this PR creates a new aggregation operator. A small code change will be required to call the aggregation operator in this PR rather than the existing `AggregateOperator`. Hope this makes sense, let us know if you have any more questions

somandal commented on pinot-query-planner/src/main/java/org/apache/calcite/rel/hint/PinotHintStrategyTable.java:
```diff
@@ -29,6 +29,9 @@
 public class PinotHintStrategyTable {
   public static final String INTERNAL_AGG_INTERMEDIATE_STAGE = "aggIntermediateStage";
   public static final String INTERNAL_AGG_FINAL_STAGE = "aggFinalStage";
+  // Hint to denote that aggregation is to be run as a single stage rather than split into an intermediate and a final
+  // stage
+  public static final String INTERNAL_IS_SINGLE_STAGE_AGG = "isSingleStageAgg";
```
@walterddr this will make more sense once you look at this PR: https://github.com/apache/pinot/pull/10846
We cannot leverage the final stage hint because we need to differentiate between the intermediate stage actually running in the v1 engine as a leaf stage vs. intermediate stage running in the v2 engine after performing operations such as JOIN (agg on top of join vs. agg in leaf)

vvivekiyer commented on pinot-common/src/main/java/org/apache/pinot/common/request/context/IdentifierContext.java:
```diff
@@ -0,0 +1,75 @@
+ * multistage query engine uses ordinals to distinctly track each identifier. So this context is set up to support both
+ * v1 and multistage engine identifiers.
+ */
+public class IdentifierContext {
```
V1 engine uses the actual column name and uses it to fetch the values.
In V2 engine, we need the ordinal info (index) to fetch the values from container rows. 

As you suggested, I did think about constructing a custom name like "$colname_$ordinal" to use in Identifier. But this would involve using an expensive string operations to extract the ordinal information from the identifier name (in `extractIntermediateValue`) and hence decided against it.  Please let me know your thoughts.

vvivekiyer commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/NewAggregateOperator.java:
```diff
@@ -0,0 +1,378 @@
+  }
+
+  public void aggregateValues(List<Object[]> container) {
+    // Convert row to columnar representation
+    Map<Integer, List<Object>> columnValuesMap = new HashMap<>();
```
Answered below.

vvivekiyer commented on pinot-core/src/main/java/org/apache/pinot/core/common/IntermediateStageBlockValSet.java:
```diff
@@ -0,0 +1,250 @@
+  private final RoaringBitmap _nullBitMap;
+  private boolean _nullBitMapSet;
+
+  public IntermediateStageBlockValSet(DataSchema.ColumnDataType columnDataType, List<Object> values) {
```
We just fetch the `Object` value from every row of container and pass it as a list of objects here. There's no boxing to Object that happens here. Am I missing something? 

Are you suggesting that when we call `DataBlockUtils.extractRows()` we can avoid the boxing to Object as we already know the type and that we can use it here in `IntermediateStageBlockValSet`? If yes, this could be a great optimization that we could pick up orthogonal to this PR and that would also address your comment about using DataBlockUtils above.

> is there any better way of handling this? converting row to columnar should be done via RowDataBlock and ColumnDataBlock abstraction?
> 
> I agree we can do the refactoring later but was wondering if there's any way we can directly extract this part into a DataBlockUtils

vvivekiyer commented on pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggFunctionQueryContext.java:
```diff
@@ -0,0 +1,62 @@
+import org.apache.pinot.core.query.request.context.QueryContext;
+
+
+/**
+ * The <code>AggFunctionQueryContext</code> class contains extracted details from QueryContext that can be used for
+ * Aggregation Functions.
+ */
+public class AggFunctionQueryContext {
```
` AggregationFunctionFactory.getAggregationFunction()` used queryContext earlier to figure out the details like - nullHandling, Limit, OrderBy.

We need this refactor for this effort because the intermediate operators are unaware of QueryContext.

vvivekiyer commented on pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunction.java:
```diff
@@ -106,6 +106,27 @@ void aggregateGroupByMV(int length, int[][] groupKeysArray, GroupByResultHolder
+   */
+  default void mergeAndUpdateResultHolder(IntermediateResult intermediateResult,
+      AggregationResultHolder aggregationResultHolder) {
+    // TODO: Remove when support for all aggregation functions is added to the Multistage engine.
```
The end state is that every aggregation function that is yet to be supported (eg: mode, covariance) will have their own implementations of `mergeAndUpdateResultHolder()`

vvivekiyer commented on pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionFactory.java:
```diff
@@ -187,17 +189,19 @@ public static AggregationFunction getAggregationFunction(FunctionContext functio
+          // TODO(Sonam): Uncomment SUM0 when merging planner changes
           case SUM:
-            return new SumAggregationFunction(firstArgument, queryContext.isNullHandlingEnabled());
+          // case SUM0:
```
The return type etc is set by the planner changes in [AggregationFunctionType.java](https://github.com/apache/pinot/pull/10846/files#diff-57508c9bebed5cdf9f9fcdbf19728615ae9f224f9e09e42de03e60f5235c7306)

vvivekiyer commented:
Hey @walterddr , sorry about the late response (I was on-call last week). @somandal  and I could do a code walk-through to explain our changes. Would that make the review process easier?

vvivekiyer commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/plan/PhysicalPlanVisitor.java:
```diff
@@ -96,10 +102,85 @@ public MultiStageOperator visitMailboxSend(MailboxSendNode node, PhysicalPlanCon
+    DataSchema inputSchema = node.getInputs().get(0).getDataSchema();
+    DataSchema resultSchema = node.getDataSchema();
+
+    // Convert aggCalls to FunctionContext and ExpressionContext that our aggregation functions understand.
```
Move this logic to convert RexExpression to FunctionContext/ExpressionContext -> NewAggregateOperator.java

vvivekiyer commented on pinot-common/src/main/java/org/apache/pinot/common/request/context/ExpressionContext.java:
```diff
@@ -74,8 +83,27 @@ public LiteralContext getLiteral(){
+    return _identifier.getName();
+  }
+
+  public int getIdentifierIndex() {
```
Can we make this change contained in v2 (in NewAggregateOperator) so that we don't need this interface change?

vvivekiyer commented on pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggregationFunctionFactory.java:
```diff
@@ -309,8 +313,8 @@ public static AggregationFunction getAggregationFunction(FunctionContext functio
           case DISTINCTAVGMV:
             return new DistinctAvgMVAggregationFunction(firstArgument);
           case DISTINCT:
-            return new DistinctAggregationFunction(arguments, queryContext.getOrderByExpressions(),
```
@Jackie-Jiang  can we clean this up so that Distinct is not modeled as an aggregation function?

vvivekiyer commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/NewAggregateOperator.java:
```diff
@@ -0,0 +1,378 @@
+      System.arraycopy(keyElements, 0, row, 0, keyElements.length);
+      for (int i = 0; i < _aggregationFunctions.length; i++) {
+        row[i + _groupSet.size()] = _aggregationFunctions[i].extractGroupByResult(_groupByResultHolders[i],
+            _groupIdMap.get(e.getKey().hashCode()));
```
Don't use hashCode because there will be collisions.

vvivekiyer replied:
Use
Key -> Object array  (Intermediate results for various aggregations)


Also look at reusing IndexTable.

vvivekiyer commented on pinot-core/src/main/java/org/apache/pinot/core/common/IntermediateStageBlockValSet.java:
```diff
@@ -0,0 +1,250 @@
+  public IntermediateStageBlockValSet(DataSchema.ColumnDataType columnDataType, List<Object> values) {
+    _dataType = columnDataType.toDataType();
+    _pinotDataType = PinotDataType.getPinotDataTypeForExecution(columnDataType);
+    _values = values;
```
Instead of extracting the list of objects here 
- store the original block here (entire block) along with an index.
- When you are actually extracting the primitive types, use the index to get the value out.

xiangfu0 commented on pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/AggFunctionQueryContext.java:
```diff
@@ -0,0 +1,62 @@
+ * Aggregation Functions.
+ */
+public class AggFunctionQueryContext {
+  private boolean _isNullHandlingEnabled;
```
Make those member variables final?

vvivekiyer commented:
- Incorporated the review comments from @Jackie-Jiang  and @walterddr 
- Updated the steps to merge this PR and the planner PR https://github.com/apache/pinot/pull/10846 in the PR Description.
- Updated the future work in the PR Description

This PR is ready to be reviewed @walterddr @Jackie-Jiang @kishoreg @xiangfu0

walterddr commented on pinot-common/src/main/java/org/apache/pinot/common/datablock/DataBlockUtils.java:
```diff
@@ -268,4 +269,381 @@ private static Object[] extractRowFromDataBlock(DataBlock dataBlock, int rowId,
+   *
+   * @return int array of values in the column
+   */
+  public static int[] extractIntRowsForColumn(DataBlock dataBlock, int columnIndex) {
```
1. name is confusing suggest
```suggestion
  public static int[] extractIntValuesForColumn(DataBlock dataBlock, int columnIndex) {
```
2. (stretch/follow-up) dataBlock can be either ROW-based or COLUMNAR-based so check dataBlock type first is needed later

walterddr commented on pinot-common/src/main/java/org/apache/pinot/common/datablock/DataBlockUtils.java:
```diff
@@ -268,4 +269,381 @@ private static Object[] extractRowFromDataBlock(DataBlock dataBlock, int rowId,
+      case INT:
+      case BOOLEAN:
+        for (int rowId = 0; rowId < numRows; rowId++) {
+          if (nullBitmap != null && nullBitmap.contains(rowId)) {
+            continue;
+          }
```
nit: consider switching from a per-row nullbitmap check to a global switch-case

walterddr commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/MultistageAggregationExecutor.java:
```diff
@@ -0,0 +1,192 @@
+    DataSchema.ColumnDataType dataType = inputDataSchema.getColumnDataType(index);
+    Preconditions.checkState(block.getType().equals(DataBlock.Type.ROW), "Datablock type is not ROW");
+    return Collections.singletonMap(expression,
+        new IntermediateStageBlockValSet(dataType, block.getDataBlock(), index));
```
`block.getDataBlock()` is lazy. e.g. if the previous block is not mailbox received, then the getDataBlock() will actually convert the List<Object[]> unserialized format into a serialized format of BaseDataBlock, then convert it back to the column value primitive array. this is very inefficient.

vvivekiyer replied:
Added a TODO as discussed offline.

walterddr replied:
yeah i forgot about the nullbitmap. so long story short we might really have to do some rethinking on this

walterddr commented on pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/MultistageGroupByExecutor.java:
```diff
@@ -0,0 +1,270 @@
+
+    DataSchema.ColumnDataType dataType = inputDataSchema.getColumnDataType(index);
+    Preconditions.checkState(block.getType().equals(DataBlock.Type.ROW), "Datablock type is not ROW");
+    return Collections.singletonMap(expression,
+        new IntermediateStageBlockValSet(dataType, block.getDataBlock(), index));
```
same here. consider which is the format we are going to use. the serialized or the unserialized.

walterddr reviewed (commented):
lgtm mostly. thank you for reformating the interface/APIs to be self contained

please kindly see the follow up comments

walterddr commented on pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/BaseBooleanAggregationFunction.java:
```diff
@@ -233,7 +233,8 @@ public Integer merge(Integer intermediateResult1, Integer intermediateResult2) {
   @Override
   public DataSchema.ColumnDataType getIntermediateResultColumnType() {
-    return DataSchema.ColumnDataType.BOOLEAN;
+    // The intermediateResult type is integer. The final result is converted to a boolean.
+    return DataSchema.ColumnDataType.INT;
```
this could cause backward-compatibility issue when upgrading. let's make sure this doesn't cause issues (where is this function being used in v1) 

in v2 as long as the expected data type conforms with calcite we are good

vvivekiyer replied:
Got it. 
Addressed this by converting Boolean objects to Integer objects in NewAggregateOperator.

walterddr commented on pinot-core/src/main/java/org/apache/pinot/core/query/aggregation/function/FourthMomentAggregationFunction.java:
```diff
@@ -30,7 +30,7 @@
 import org.apache.pinot.segment.spi.AggregationFunctionType;
 
-
+// TODO: Add null handling.
```
null support is not there in v1 anyway (for most of agg functions)

walterddr reviewed (approved):
lgtm. thank you for addressing the interface and interaction changes. 
i believe the PR is in a mergeable state where the result of the TODO can be address separately.

walterddr commented:
merging this one first. and continue if any follow up is needed on #10846
